billing-client/client.go

```diff
@@ -0,0 +1,157 @@
+	if err != nil {
+		return nil, err
+	}
+	intPort, err := strconv.Atoi(port)
+	if err != nil {
+		return nil, err
+	}
+	logger, err := fluent.New(fluent.Config{
+		FluentPort:    intPort,
+		FluentHost:    host,
+		AsyncConnect:  true,
+		MaxRetry:      -1,
+		MarshalAsJSON: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		stop:   make(chan struct{}),
+		events: make(chan Event, maxBufferedEvents),
+		logger: logger,
+	}
+	c.wg.Add(1)
+	go c.loop()
+	return c, nil
+}
+
+// AddAmounts writes unit increments into the billing system. If the call does
+// not complete (due to a crash, etc), then data may or may not have been
+// written successfully.
+//
+// Requests with the same `uniqueKey` can be retried indefinitely until they
+// succeed, and the results will be deduped.
+//
+// `uniqueKey` must be set, and not blank. If in doubt, generate a uuid and set
+// that as the uniqueKey.
+//
+// `internalInstanceID`, is *not* the external instance ID (e.g. "fluffy-bunny-47"), it is the numeric internal instance ID (e.g. "1234").
+//
+// `timestamp` is used to determine which time bucket the usage occured in, it is included so that the result is independent of how long processing takes.
+//
+// `amounts` is a map with all the various amounts we wish to charge the user for.
+//
+// `metadata` is a general dumping ground for other metadata you may wish to include for auditability. In general, be careful about the size of data put here. Prefer including a pointer over whole data. For example, include a report id or s3 address instead of the information in the report.
+func (c *Client) AddAmounts(uniqueKey, internalInstanceID string, timestamp time.Time, amounts Amounts, metadata map[string]string) error {
+	if uniqueKey == "" {
+		return fmt.Errorf("billing units uniqueKey cannot be blank")
+	}
+
+	e := Event{
+		UniqueKey:          uniqueKey,
+		InternalInstanceID: internalInstanceID,
+		OccurredAt:         timestamp,
+		Amounts:            amounts,
+		Metadata:           metadata,
+	}
+
+	select {
+	case <-c.stop:
+		// TODO: Should this be a panic? We just threw away billing data!
+		return fmt.Errorf("Stopping, discarding event: %v", e)
+	default:
+	}
+
+	select {
+	case c.events <- e: // Put event in the channel unless it is full
+		return nil
+	default:
+		// full
+	}
+	// TODO: Big angry log message here, cause we're throwing away billing data
+	return fmt.Errorf("Reached billing event buffer limit (%d), discarding event: %v", maxBufferedEvents, e)
+}
+
+func (c *Client) loop() {
+	defer c.wg.Done()
+	for done := false; !done; {
+		select {
+		case event := <-c.events:
+			c.post(event)
+		case <-c.stop:
+			done = true
+		}
+	}
+
+	// flush remaining events
+	for done := false; !done; {
+		select {
+		case event := <-c.events:
+			c.post(event)
+		default:
+			done = true
+		}
+	}
+}
+
+func (c *Client) post(e Event) error {
+	for {
+		select {
+		case <-c.stop:
+			return fmt.Errorf("Billing: failed to log event: %v", e)
```

Comment:
Won't this interact badly with the flushing in `loop` - ie won't it always fail to flush anything after being stopped?